model/lifecycle.go

```diff
@@ -1080,25 +1083,131 @@ func setNumDeps(tasks []*task.Task) {
 	for i, task := range tasks {
 		idToTask[task.Id] = tasks[i]
 	}
-
-	for _, task := range tasks {
+	deduplicatedTasks := []*task.Task{}
+	for _, task := range idToTask {
+		task.NumDependents = 0
+		deduplicatedTasks = append(deduplicatedTasks, task)
+	}
+	for _, task := range deduplicatedTasks {
 		// Recursively find all tasks that task depends on and increments their NumDependents field
 		setNumDepsRec(task, idToTask, make(map[string]bool))
 	}
 }
 
 // setNumDepsRec recursively finds all tasks that task depends on and increments their NumDependents field.
 // tasks not in idToTasks are not affected.
-func setNumDepsRec(task *task.Task, idToTasks map[string]*task.Task, seen map[string]bool) {
-	for _, dep := range task.DependsOn {
+func setNumDepsRec(t *task.Task, idToTasks map[string]*task.Task, seen map[string]bool) {
+	for _, dep := range t.DependsOn {
 		// Check whether this dependency is included in the tasks we're currently creating
-		if depTask, ok := idToTasks[dep.TaskId]; ok {
-			if !seen[depTask.Id] {
-				seen[depTask.Id] = true
-				depTask.NumDependents = depTask.NumDependents + 1
-				setNumDepsRec(depTask, idToTasks, seen)
-			}
+		depTask, ok := idToTasks[dep.TaskId]
+		if !ok {
+			// it's possible to get in to this block if tasks depend on others outside of the task's
+			// version, such as with commit queue merges. This scenario is currently not handled here
+			// because commit queue merges are all linked lists, and this block would do nothing
+			continue
 		}
+		if !seen[depTask.Id] {
+			seen[depTask.Id] = true
+			depTask.NumDependents = depTask.NumDependents + 1
+			setNumDepsRec(depTask, idToTasks, seen)
+		}
+	}
+}
+
+func RecomputeNumDependents(t task.Task) error {
```

Comment:
We only use NumDependents for the scheduler, right? Is it so important that we have the number of dependencies exactly right? This is an involved function.